control_plane/attachment_service/src/service.rs

```diff
@@ -742,42 +745,275 @@ impl Service {
             (waiters, response_shards)
         };
 
-        let deadline = Instant::now().checked_add(Duration::from_secs(5)).unwrap();
+        self.await_waiters(waiters).await?;
+
+        Ok(TenantCreateResponse {
+            shards: response_shards,
+        })
+    }
+
+    /// Helper for functions that reconcile a number of shards, and would like to do a timeout-bounded
+    /// wait for reconciliation to complete before responding.
+    async fn await_waiters(
+        &self,
+        waiters: Vec<ReconcilerWaiter>,
+    ) -> Result<(), ReconcileWaitError> {
+        let deadline = Instant::now().checked_add(Duration::from_secs(30)).unwrap();
         for waiter in waiters {
             let timeout = deadline.duration_since(Instant::now());
             waiter.wait_timeout(timeout).await?;
         }
-        Ok(TenantCreateResponse {
-            shards: response_shards,
-        })
+
+        Ok(())
     }
 
-    pub(crate) async fn tenant_timeline_create(
+    /// This API is used by the cloud control plane to do coarse-grained control of tenants:
+    /// - Call with mode Attached* to upsert the tenant.
+    /// - Call with mode Detached to switch to PolicyMode::Detached
+    ///
+    /// In future, calling with mode Secondary may switch to a detach-lite mode in which a tenant only has
+    /// secondary locations.
+    pub(crate) async fn tenant_location_config(
         &self,
         tenant_id: TenantId,
-        mut create_req: TimelineCreateRequest,
-    ) -> Result<TimelineInfo, ApiError> {
-        let mut timeline_info = None;
+        req: TenantLocationConfigRequest,
+    ) -> Result<TenantLocationConfigResponse, ApiError> {
+        if req.tenant_id.shard_count.0 > 1 {
+            return Err(ApiError::BadRequest(anyhow::anyhow!(
+                "This API is for importing single-sharded or unsharded tenants"
+            )));
+        }
 
-        let ensure_waiters = {
-            let locked = self.inner.write().unwrap();
+        let mut waiters = Vec::new();
+        let mut result = TenantLocationConfigResponse { shards: Vec::new() };
+        let maybe_create = {
+            let mut locked = self.inner.write().unwrap();
+            let result_tx = locked.result_tx.clone();
+            let compute_hook = locked.compute_hook.clone();
+            let pageservers = locked.nodes.clone();
+
+            let mut scheduler = Scheduler::new(&locked.tenants, &locked.nodes);
+
+            // Maybe we have existing shards
+            let mut create = true;
+            for (shard_id, shard) in locked
+                .tenants
+                .range_mut(TenantShardId::tenant_range(tenant_id))
+            {
+                // Saw an existing shard: this is not a creation
+                create = false;
+
```

Comment:
So, in this service's data model, we don't have a `Tenant`, but only `TenantShard`?
Feels wrong to me, e.g., this `create` hack.

Also, we definitely have some Tenant-scoped information, e.g., when control plane calls LocationConfigMode::Detached for a sharded tenant, what that means is "desired state: tenant doesn't consume resources on pageservers", not "update each tenant shard's placement policy to detached".

Reply:
It's true that there is some state that is logically per-tenant, which we duplicate across its shards (e.g. the placement policy).  It's a useful simplification in some places, to avoid having a 2-level hierarchy of objects, but in other places (like here), it results in a little extra complexity to express the "any shards?" concept rather than the singular "does tenant X exist" concept.  There's a similar tradeoff with e.g. worries of shards ending up inconsistent with one another: we have to get that right, but if we had a hierarchy we would have to worry about making sure that we never had a tenant with zero shards, or a shard with no parent tenant, etc.

From a high level architecture point of view, we can think of the control plane as owning the Tenant (the logical concept the user sees), and this service owning tenant shards (the physical concept that pageservers see).

If we moved to a hierarchy of `Tenant`->`TenantShard`, we would still have loops in many places: for example when changing the policy on a tenant, one would still have to loop over shards to call schedule() and maybe_reconcile() on them.

I'm not super-committed to this, but for now I find it a useful simplification to just have two collections of objects (shards and nodes)